### PR TITLE
[aggregator/ckey] implement a faster hash to reduce CPU and RAM usage 

### PR DESCRIPTION
**Pull request by remeh**
### What does this PR do?

The highest CPU usage in the CPU profiles of highly loaded DogStatsD server was
the sorting of the metric tags in order to generate reproducible context keys.

This commit implements a hash which does not need to sort the tags to generate
reproducible context keys: it relies on XORing parts of the hash together instead
of building a bytes buffer containing all parts before hashing this whole buffer.
In order for this to work, a fast deduplication algorithm had to be written since
duplicated tags would make the hash computation wrong because of the XOR usage.

Using this implementation, the DogStatsD server will use less CPU to process more
metrics, and thus, it will use less RAM since it won't have to buffer as much as
before.

Micro-benchmarks of the KeyGenerator, `7.30.0-dev` against this branch:

```
KeyGeneration/1-tags-8    23.6ns ± 1%  21.7ns ± 1%   -8.03%  (p=0.008 n=5+5)
KeyGeneration/2-tags-8    33.2ns ± 1%  29.9ns ± 1%  -10.04%  (p=0.008 n=5+5)
KeyGeneration/4-tags-8    55.2ns ± 0%  48.1ns ± 0%  -12.79%  (p=0.008 n=5+5)
KeyGeneration/8-tags-8     102ns ± 1%    91ns ± 0%  -10.79%  (p=0.016 n=5+4)
KeyGeneration/16-tags-8    203ns ± 2%   189ns ± 4%   -7.11%  (p=0.008 n=5+5)
KeyGeneration/32-tags-8    397ns ± 1%   314ns ± 2%  -20.75%  (p=0.008 n=5+5)
KeyGeneration/64-tags-8   2.61µs ± 1%  0.64µs ± 5%  -75.67%  (p=0.008 n=5+5)
KeyGeneration/128-tags-8  6.23µs ± 7%  1.24µs ± 2%  -80.12%  (p=0.008 n=5+5)
```

Here are different benchmarks results, where DogStatsD servers are hit with
300k metrics/second, with 10, 15 and 20 client tags added to the metrics
(the benchmarks are configured with "High cardinality" discovery tags, meaning
that the metrics end up with more than 50 tags approximately).

**TODO(remy) update these graphs**

PR results are in red/orange in those graphs, purple lines are `7.29.0`: 

image
image
image

### Motivation

Improve performances of the DogStatsD server to lower its CPU and RAM usage.

### Additional Notes

I've decided to continue to sort+unique the tags for Service Checks and (old) Events since their throughput isn't as high as DogStatsD samples.

AFAIK, mathematically, nothing has changed about the amount of collisions possible with this context key implementation, still, I've created a tool to naively / practically verify this in https://github.com/DataDog/datadog-agent/pull/8526 and this implementation has been through it multiple times successfully (i.e. tested with more than 50 millions different generated metrics).

This is a v2 of https://github.com/DataDog/datadog-agent/pull/8529 which had an unwanted behavior: sending tags unduplicated to the backend.

### Describe how to test your changes

Validate that custom metrics received by DogStatsD are still available and display sane values.

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [N/A] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [N/A] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

### CODE REVIEW

File: pkg/dogstatsd/server.go [hunk 1]
```diff
@@ -691,13 +691,18 @@ func (s *Server) Stop() {
 	s.Started = false
 }
 
+// storeMetricStats stores stats on the given metric sample.
+// It is storing the stats of the sample as they are received: it means that if
+// a metric is received with duplicated tags, it will be stored this way here.
+// It can help troubleshooting clients with bad behaviors but it is different that
+// what will be outputted by the aggregator, which takes care of deduping the tags
+// to aggregate the metrics.
```
Comment (djmitche) [inline pkg/dogstatsd/server.go:699]: ++ for godocs :)

File: pkg/aggregator/ckey/key.go [hunk 2]
```diff
@@ -27,49 +23,106 @@ import (
...
+	OUTER:
+		for i := range tags {
+			h := murmur3.StringSum64(tags[i])
+			for j := 0; j < g.idx; j++ {
+				if g.seen[j] == h {
+					continue OUTER // we do not want to xor multiple times the same tag
+				}
+			}
+			g.intb = g.intb ^ h
+			g.seen[g.idx] = h
```
Comment (djmitche) [inline pkg/aggregator/ckey/key.go:128]: This fails if `g.idx > 512`, which occurs when there are >512 tags

File: pkg/aggregator/ckey/key.go [hunk 3]
```diff
@@ -27,49 +23,119 @@ import (
...
 
-	g.buf = append(g.buf, name...)
-	g.buf = append(g.buf, ',')
-	for i := 0; i < len(tags); i++ {
-		g.buf = append(g.buf, tags[i]...)
-		g.buf = append(g.buf, ',')
+	// there is two implementations used here to deduplicate the tags depending on how
+	// many tags we have to process:
+	//   -  16 < n < hashSetSize:	we use a hashset of `hashSetSize` values.
+	//   -  n < 16 or n > hashSetSize: we use a simple for loops, which is faster than
```
Comment (djmitche) [inline pkg/aggregator/ckey/key.go:77]:
```suggestion
	//   -  n < 16 or n > hashSetSize: we use a simple for loop, which is faster than
```

File: pkg/aggregator/ckey/key_test.go [hunk 4]
```diff
@@ -49,6 +49,57 @@ func TestCompare(t *testing.T) {
...
+	// add a duplicated tag
+	tags = append(tags, "key:value", "foo")
+	key3, _ := generator.Generate(name, hostname, tags)
+	assert.Equal(key, key3, "duplicated tags should not matter")
+
+	// and now, completely change of the tag, the generated key should NOT be the same
+	tags[2] = "another:tag"
+	key4, _ := generator.Generate(name, hostname, tags)
+	assert.NotEqual(key, key4, "tags content should matter")
+}
```
Comment (djmitche) [inline pkg/aggregator/ckey/key_test.go:76]: I added another test to give a bit more iterations to this testing.  It doesn't necessarily hit all of the corner cases (hash collisions, zero hashes, etc.), but it did find the issue with >512 tags.

```suggestion
}

func TestTagsAreDedupedWhileGeneratingCKey(t *testing.T) {
        withSizeAndSeed := func(size, iterations int, seed int64) func(*testing.T) {
                return func(t *testing.T) {
                        assert := assert.New(t)
                        r := rand.New(rand.NewSource(seed))

                        name := "metrics.to.test.hashing"
                        hostname := "hostname.localhost"
                        tags := genTags(size)

                        generator := NewKeyGenerator()
                        expKey, _ := generator.Generate(name, hostname, tags)

                        for i := 0; i < iterations; i++ {
                                r.Shuffle(size, func(i, j int) { tags[i], tags[j] = tags[j], tags[i] })
                                key, _ := generator.Generate(name, hostname, tags)
                                assert.Equal(expKey, key, "order of tags should not matter")
                        }
                }
        }
        t.Run("smallish", withSizeAndSeed(10, 200, 0x398192f0a9c0))
        t.Run("bigger", withSizeAndSeed(50, 100, 0x398192f0a9c0))
        t.Run("huge", withSizeAndSeed(600, 10, 0x398192f0a9c0))
}
```

**Review by djmitche — COMMENTED** [review]
I only noticed this was a draft after making all these comments, so my apologies if many/all of these were things you were going to fix anyway :)

**Comment by remeh** [comment]
Is part of #8822
